README
This is version 2.11b of ``noweb'', a simple, extensible literate-
programming tool.  noweb is available by anonymous ftp from
ftp.eecs.harvard.edu:pub/nr/noweb and by HTTP from
http://www.eecs.harvard.edu/nr/noweb/dist.  The ftp site is mirrored
by the Comprehensive TeX Archive Network, in directory web/noweb.
The home page at http://www.eecs.harvard.edu/~nr/noweb contains much
helpful information.

To build noweb, change to the `src' directory and follow the
instructions in the file INSTALL.


WHAT'S NEW IN VERSION 2.11b

Version 2.11b is a bugfix version: Fixed the noroff script to work
with current GNU roff, and fixed noweave -x -troff at least to the
extent that noroff doesn't choke on its output.


INTRODUCTION --- WHAT IS NOWEB, ANYWAY?

noweb is designed to meet the needs of literate programmers while
remaining as simple as possible.  Its primary advantages are
simplicity, extensibility, and language-independence.  noweb uses 5
control sequences to WEB's 27.  The noweb manual is only 3 pages; an
additional page explains how to customize its LaTeX output.  noweb
works ``out of the box'' with any programming language, and supports
TeX, latex, and HTML back ends.  A back end to support full hypertext
or indexing takes about 250 lines; a simpler one can be written in 40
lines of awk.  Noweb does not prettyprint natively, but supports
prettyprinting through such third-party filters as pretzel and dpp.

noweb has been distributed free of charge for ten years, and it is
one of the world's most widely used literate-programming tools.  It
has been used for hundreds of thousands of lines of code in such
languages as awk, C, C++, Icon, Modula-3, PAL, perl, Promela, and
Standard ML.

If you already know you want to use noweb, you need only install it
and read the manual page.  If you're just curious about noweb, read
the paper that appeared in the September 1994 issue of IEEE Software.
(If you can't get Software, send me a postcard and I'll send you a
reprint.)  A nice, brief tutorial appeared in the October 1997 issue
of Linux Journal (but beware that chunk syntax is <<name>> and not <<name>).

If you're brand new to literate programming, check out the FAQ for the
USENET newsgroup comp.programming.literate.   There are also some
resources available through the noweb home page:
   http://www.eecs.harvard.edu/~nr/noweb


WHAT YOU GET IN THIS DISTRIBUTION

This distribution contains the following directories:
  binaries	Pre-built distributions containing binaries and documentation
  contrib	software contributed by noweb users 
  examples	sample noweb filters and programs in different languages
  src		Source code and documentation (including FAQ) for noweb
Where appropriate, these directories have README files of their own.
BINARY DISTRIBUTIONS MAY NOT ALWAYS BE UP-TO-DATE, especially DOS
distributions.
  

EXTENSIBILITY

noweb provides extensibility by using the Unix toolkit philosophy.
The ``noweb,'' ``notangle,'' and ``noweave'' commands are built from
pieces, which are then assembled in pipelines using shell scripts.
The pieces include:

  markup	convert noweb file from human syntax to tool syntax
  unmarkup	inverse of markup
  totex		convert from tool syntax to TeX/latex markup
  tohtml	convert from tool syntax to HTML markup
  nt		`tangle' the tool form of the noweb file
  mnt		discover roots, then act like nt
  noidx		insert indexing and cross-reference information
  finduses	finds uses of identifiers

These pieces are combined by the scripts in the src/shell directory to
provide more than just weaving and tangling:

  noweb		analog of nuweb
  notangle	analog of TANGLE
  noweave	analog of WEAVE
  nountangle	tangle, but keep interleaved documentation in comments
  noroots	print names of all root chunks in a noweb file
  nocount	count number of lines of code and documentation.
  nodefs	extract defined identifiers for noweave -indexfrom
  noindex	build an external index for multi-file documents


WEAVING --- A TAR PIT

The worst aspect of literate programming is the enormous amount of
time wasted wrangling over what ``woven'' output should look like.
Although noweb does no prettyprinting, it is not entirely immune---
several people have complained about noweave's output or have sent me
changes that add more options to noweave.  I resisted for years, but
with version 2.5 I finally succumbed.  I let the number of options to
noweave double, and I have provided for too many options and hooks for
customizing the latex output.  I won't let it happen again.

noweb doesn't try to be all things to all programmers, but it is very
easy to change.  If you don't like noweave's formatting, you can read
tex/support.nw to learn how to customize it; look for the words
``style hook.''  (Reading noweb.sty directly is not recommended.)  For
simple formatting, it might be easier to throw away noweave and make
your own.  To help you get started, the shell directory contains
noweave.simple, a simplified version of noweave that Dave Hanson
created for use with C programs (it can't handle code with @ signs).
The Noweb Hacker's Guide (xdoc/guide.tex) explains the intermediate
language that noweb uses to represent literate programs.

The intermediate language makes it possible to extend noweave with a
language-dependent prettyprinter, as shown by contributions of an Icon
prettyprinter by Kostas Oikonomou and a guarded-command prettyprinter
by Conrado Martinez-Parra.  (I haven't written a prettyprinter myself
because my experience with Spider left me thinking that prettyprinting
is far more trouble than it's worth.)  Further contributions of
prettyprinters are welcome.

noweb comes with two cross-referencers for use with noweave.  The
standard one is written in Icon, because Icon offers excellent
functionality and performance.  Because Icon is not available on all
platforms, I profide an alternative, but inferior cross-referencer
written in awk.  See the INSTALL file for details.

Cross-referencing makes formatting even more of a tar pit; the
cross-referencer itself takes about 300 lines, and extensive LaTeX
support is also required.  I haven't made the attempt to write
cross-reference code for plain TeX.  Anyone who has ideas for reducing
the number of options or for other ways to restore sanity to the
situation is urged to write to nr@eecs.harvard.edu.


INDEX AND IDENTIFIER CROSS-REFERENCE

Noweb creates identifier cross-references so that you can click on an
identifier and jump to its definition (if you're using printed LaTeX
output, a footnote gives you the page number of the definition).  
To noweb, any string of nonwhite characters can be an identifier.  
A human being or a language-dependent tool must mark definitions of
identifiers; noweb finds the uses using a language-independent
algorithm.  The algorithm relies on an idea taken from the lexical
conventions of Standard ML.  Characters are divided into three
classes: alphanumerics, symbols, and delimiters.  If an identifier
begins with an alphanumeric, it must be delimited on the left by a
symbol or a delimiter.  If it begins with a symbol, it must be
delimited on the left by an alphanumeric or a delimiter.  If it begins
with a delimiter, there are no restrictions on the character
immediately to the left.  Similar rules apply on the right-hand side.
The default classifications are chosen to make sense for commonly used
programming languages, so that noweb will not recognize `zip' when it
sees `zippy', or `++' when it sees `++:='.  This trick works
surprisingly well, but it does not prevent noweb from spotting
identifiers in comments or string literals.

The basic assumption in noweb is that a language-dependent filter will
identify definitions automatically.  Filters for Icon, TeX, and yacc
all take about 30 lines of Icon code and are included in the noweb
distribution; if you write a filter of your own, you can put it in the
$LIB directory with a name like `autodefs.pascal'.  Try `noweave
-showautodefs' for a complete list of such filters. 

If the automatic filter does not work well for you, or if there is no
filter available for your language, I recommend that you mark
definitions using backticks (`) in your source code, and use
`-filter btdefn' with both noweave and notangle.

noweave -index works well for short programs, but nodefs, noindex, and
noweave -indexfrom are there for large multi-file programs.  See the
noindex man page for details.


NOTES

  src/xdoc/techrep.* contains an early, almost unrecognizable version
  of a paper describing noweb that appeared in IEEE Software in
  September, 1994.  You are probably better off writing me for a
  reprint of the Software paper.  (Send a postcard!)

  The Noweb Hacker's Guide, which appears in src/xdoc/guide.*,
  documents the representation of noweb files that is used by the
  noweb tools, in case you want to write any tools of your own.
  Simple tools (e.g. count the number of lines of interleaved
  documentation) are trivial.  If you write any tools, or you want
  tools written (e.g. prettyprinters, index generators), let me know.

  The icon directory contains Icon versions of most pipeline
  stages.  If you want to adapt noweb to work with a text processor
  other than TeX or latex, they might provide a better starting point.
  Perhaps the whole system should have been written in Icon from the
  beginning, but I'm not going to do it over.  Icon is available by
  anonymous ftp from cs.arizona.edu.

  Phil Bewig has adapted noweb to troff, and as soon as we get the
  portability bugs ironed out, I'll include that code in the noweb
  distribution.

  
  Thanks to Preston Briggs for the Aho-Corasick recognizer, and for
		helpful discussions.
  Thanks to Dave Hanson for cpif.
  Thanks to Dave Love for LaTeX wizardry.
  Thanks to Joseph Reynolds for prodding me to fix [[...]].
  Thanks to Bill Trost for the original HTML back end.
  Thanks to Philip Miller and Lee Wittenberg for DOS binaries.
  Thanks to Paolo Ciccone for the Win32 binaries.

Send comments or questions to nr@eecs.harvard.edu.  

I enjoy hearing from noweb users; if you have enjoyed noweb, why not
send me a local postcard for my collection?  My address is:

  Norman Ramsey
  Division of Engineering and Applied Sciences
  Maxwell Dworkin 231
  Harvard University
  33 Oxford Street
  Cambridge, MA 02138
               USA


COPYRIGHT

Noweb is copyright 1989-2006 by Norman Ramsey.  All rights reserved.

You may use and distribute noweb for any purpose, for free.  You may
modify noweb and create derived works, provided you retain the
copyright notice, but the result may not be called noweb without my
written consent.  You may not sell noweb itself, but you may do
anything you like with programs created with noweb.


CHANGES
CHANGES FOR VERSION 2.11b - released 12 Jun 2006
Make noroff use -mm by default (Debian 218050)
Fixed flagrant errors in noweave -x -troff (but bad formatting remains)

CHANGES FOR VERSION 2.11a - released 20 May 2006
Clarify instructions for $PATH and $MANPATH (B Pierce)
Create examples/Makefile.awk (B Pierce)
Clarify documentation: unescaped << and >> not OK in docs chunk (R Hemmecke)
Add -lua option to docs2comments filter

CHANGES FOR VERSION 2.11 - released 6 April 2006
Braces did not render correctly in [[<<$\{a_i\}$>>]] (Albert Reiner)
From Kevin Quick, Windows hints and a new NOTANGLE.BAT
Generate postscript with -P cmz for better PDF (Francky Leyn)
Clarify for noweave -delay that the first doco chunk is before the first @
extended xchunks with -summary argument
Line in a code chunk opening with unclosed, unescaped << is OK (Ralf Hemmecke)

UNDO ME!  Make docs2comments filter tolerate embedded @xref stuff..
Added Windows XP help to FAQ (David Lemmon)
Noidx now calculates labels for a code chunk using the number of preceding
  definitions from the same @file, not the overall number
Change wording for \noweboptions{german} (Christian Lindig)
Need to insert \leavevmode in \sublabel so that new \@bsphack will work right
From Yann Dirson, new filters in contrib/ydirson.
From Fr'ed'eric Lin'e, some improvements in French translation (apologies to Moli`ere)

CHANGES FOR VERSION 2.10c - released 17 Jul 2002

Bug in noidx.nw caused tag NWD1 etc to be used in multiple files.
Cleaned up alignment tags in tables emitted by l2h.nw (Federico Di Gregorio)
Fixed some bugs in the HTML markup for the hacker's guide
Goes back to iconc, except for noidx.nw, which can't be compiled because
  of the 32-bit polynomial arithmetic

CHANGES FOR VERSION 2.10b - released 15 Jul 2002

Bug in xchunks.nw, around line 100 -- db never initialized (Stephen Edwards)
In local identifier cross-reference, enable breaks between pages 
   of multiply defined identifiers (option breakdefs)
Create a secret \noweboptions{breakcode} for those who want their
   code chunks broken across page boundaries
German is `Benutzt' not `Benutztt' (Johannes Huesing)
No longer suggests iconc; compiling with icont is now the default.

CHANGES FOR VERSION 2.10a - released 30 Mar 2001

Fixed problem in src/Makefile -- FAQ should be built by `source' not `all'

CHANGES FOR VERSION 2.10 - released 16 Jun 2000, but not officially

Added -matlab to nountangle (Uwe Brauer)
Added \nwblindhyperanchor to \weblabel (Andrew Swann)
Changed one more 'char c' to 'unsigned char c' in recognize.nw (Gilles ANDRE)
Change noweb.sty to support more than 26 code chunks on a page (F. Leyn)
l2h skips optional arguments with \includegraphics
fixed bugs in htmltoc having to do with nested anchors on one line
Added capacity for two more filters to noweave script
Fixed bug in autodefs.sml (including spaces in symbol chars)
Added autodefs.mmix
Added autodefs.asdl
Scripts for Plan 9 rc (Russ Cox)
Tinkered with noidx sortkey
Added \mbox{} to \nwnewline, so that spaces won't be eaten (John Cummings)
Added xchunks filter, per request of Mike Smith & Glenn Holloway
Replaced FORTH uniqueid heuristic with 32-bit CRC
Added pipedocs filter to pipe documentation chunks through commands
Added h2a (convert html to ascii) for use with pipedoces
Added docs2comments filter, with pipedocs and h2a, to replace nountangle
Added nwmtime filter; helps derive Zephyr web pages
Moved the official FAQ to html
Added multiple-language support to noweb.sty (Miguel Filgueiras)
notangle -L initial newline is a bug and will never be fixed (Harald Soleng)
Added \noweboptions{noxref}, intended for temporary use inside documents.

CHANGES FOR VERSION 2.9a - released 16 Feb 1999

Fixed bug in htmltoc that left multiple toc anchors


CHANGES FOR VERSION 2.9

Good support for hypertext using the `hyperref' package
noweave now finds uses in quoted code, the better to exploit hypertext
added support for emacs font-lock highlighting as comments to noweb-mode (Adnan Yaqub)
Document common -L options on notangle man page (Brett Presnell)
Fixed incompatibility with verbatim package (Rafael Laboissiere)
added contrib/leyn from Francky Leyn
noweb-mode.el no longer treats @x as chunk end for non-white x (Chris Milner)
Don't use signed char to index array of 256 in recognize.nw (Michael Herrmann)
Added explanation for WNT to FAQ.
Support \ quoting in argument to elide
Support multiple -R options on the same notangle command line.
Change tohtml so that noweave -latex+html doesn't protect anchor texts
   in a rawhtml environment (???)
Added to /usr/doc/noweb for Slackware binary distribution
Slackware binary distribution now includes emacs noweb mode.
Updated internal docs in support.nw to explain the great \nwlbrace trouble
Use gcc -ansi -pedantic in Makefiles.
Remove multicols environment from examples/wc.nw (Yotam Medini)
Change noweave and totex to support subscripting identifiers used in quoted code
Fixed bug in l2h (bad conversion of tables).
Make l2h of \section use CPS (for newline between \section and {)
Added letenv to l2h, and changed verbatim to use environment name as terminator


CHANGES FOR VERSION 2.8a

noweave complains if you use -x with -index or -indexfrom
installation now includes noweb-mode.el
noroots no longer treats uses in quoted code as real uses (David Bruce)
file INSTALL now notes that 'awkname' is always needed (Robert Solovay)
file INSTALL now notes that texhash(1) may be needed (Robert Solovay)
mnt (noweb script) now complains about chunks named <<*>> (Robert Solovay)
lots of rcs identifiers and names everywhere
all man pages should carry version numbers
Use Stephen Parker's htmltoc, with many bug fixes and improvements.


CHANGES FOR VERSION 2.8

fixed unbalanced brace in support.nw for nwmac.tex (Hartmut Henkel)
changed notangle script so notangle -L preserves tabs (Saul London, NR)
changed uniqueid in noidx to avoid everything except letters, digits,

src/c/mnt.c
#line 9 "mnt.nw"
static char rcsid[] = "$Id: mnt.nw,v 1.19 2006/06/12 21:03:53 nr Exp nr $";
static char rcsname[] = "$Name: v2_11b $";
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <ctype.h>
#include "modules.h"
#include "modtrees.h"
#include "notangle.h"
#include "errors.h"
#include "columns.h"
#include "strsave.h"

#line 52 "mnt.nw"
void add_uses_to_usecounts(Module mp);
void emit_if_unused_and_conforming(Module mp);
#line 86 "mnt.nw"
static void emitfile(char *modname);
#line 180 "mnt.nw"
#ifdef TEMPNAM
extern char *tempnam (const char *dir, const char *pfx);        /* temp file in dir */
#else
#define tempnam(DIR,PFX) (strsave(tmpnam(NULL)))
#endif

#line 25 "mnt.nw"
#define Clocformat "#line %L \"%F\"%N"
static char *locformat = Clocformat;

main(int argc, char **argv) {
    int i;

    tabsize = 0;  /* default for nt is not to use tabs */

    progname = argv[0];
    finalstage = 1;
    
#line 46 "mnt.nw"
read_defs(stdin);
apply_each_module(remove_final_newline);
#line 36 "mnt.nw"
    for (i=1; i<argc; i++) 
      switch (*argv[i]) {
        case '-': 
#line 152 "mnt.nw"
    switch (*++argv[i]) {
        case 'a':
            if (strcmp(argv[i], "all"))
                errormsg(Warning, "Ignoring unknown option -%s", argv[i]);
            else {
#line 49 "mnt.nw"
apply_each_module(add_uses_to_usecounts);
apply_each_module(emit_if_unused_and_conforming);
#line 156 "mnt.nw"
                                                    }
            break;
        case 't': /* set tab size or turn off */
            if (isdigit(argv[i][1]))
                tabsize = atoi(argv[i]+1);
            else if (argv[i][1]==0)
                tabsize = 0;            /* no tabs */
            else 
                errormsg(Error, "%s: ill-formed option %s\n", argv[0], argv[i]);
            break;          
        case 'L': /* have a #line number format */
            locformat = argv[i] + 1;
            if (!*locformat) locformat = Clocformat;
            break;
        default:
            errormsg(Warning, "Ignoring unknown option -%s", argv[i]);
     }
#line 38 "mnt.nw"
                                                                   break;
        default:  emitfile(argv[i]);                               break;
      }
    nowebexit(NULL);
    return errorlevel;          /* slay warning */
}
#line 55 "mnt.nw"
void add_uses_to_usecounts(Module mp) {
    Module used;
    struct modpart *p;
    for (p=mp->head; p!=NULL; p=p->next)
        if (p->ptype == MODULE) {
            used = lookup(p->contents);
            if (used != NULL)
                used->usecount++;
        }
}
#line 71 "mnt.nw"
void emit_if_unused_and_conforming(Module mp) {
    char *index;
    if (mp->usecount == 0 && strpbrk(mp->name, " \n\t\v\r\f") == NULL)
        if (index = strpbrk(mp->name, "[](){}!$&<>*?;|^`'\\\""),
            index == NULL || index[0] == '*' && index[1] == 0)
	    if (index == mp->name)
	        errormsg(Error, "<<*>> is not a good chunk name for noweb; "
	                        "use notangle instead");
	    else
	        emitfile(mp->name);
        else 
            errormsg(Error, "<<%s>> cannot be an output chunk; "
                            "it contains a metacharacter", mp->name);
}
#line 88 "mnt.nw"
static void emitfile(char *modname) { 
  Module root = lookup(modname);
  char *tempname = tempnam(".", 0);
  FILE *fp;
  char *lfmt, *filename;
  
#line 108 "mnt.nw"
{ int n = strlen(modname) - 1;
  if (n >= 0 && modname[n] == '*') {
    lfmt = locformat;
    filename = strsave(modname);
    filename[n] = 0;
  } else {
    lfmt = "";
    filename = modname;
  }
}
#line 94 "mnt.nw"
  
#line 147 "mnt.nw"
if (root == NULL) {
  errormsg(Error, "Chunk <<%s>> is undefined", filename);
  return;
}
#line 95 "mnt.nw"
  fp = fopen(tempname, "w");
  if (fp == NULL) errormsg(Fatal, "Can't open temporary file %s", tempname);
  
#line 119 "mnt.nw"
resetloc();
(void) expand(root, 0, 0, 0, lfmt, fp);
putc('\n', fp);
fclose(fp);
#line 98 "mnt.nw"
  
#line 125 "mnt.nw"
{ FILE *dest, *tmp;
  dest = fopen(filename, "r");
  if (dest != NULL) {
    int x, y;
    tmp = fopen(tempname, "r");
    assert(tmp);
    do { 
      x = getc(tmp);
      y = getc(dest);
    } while (x == y && x != EOF);
    fclose(tmp);
    fclose(dest);
    if (x == y) {
      remove(tempname);
      return;
    }
  }
}
#line 99 "mnt.nw"
  remove(filename);
  if (rename(tempname, filename) != 0) { /* different file systems? (may have to copy) */
    FILE *fp = fopen(filename, "w");
    if (fp == NULL) {remove(tempname); 
#line 144 "mnt.nw"
errormsg(Error, "Can't open output file %s", filename);
return;
#line 102 "mnt.nw"
                                                                                     }
    
#line 119 "mnt.nw"
resetloc();
(void) expand(root, 0, 0, 0, lfmt, fp);
putc('\n', fp);
fclose(fp);
#line 104 "mnt.nw"
    remove(tempname);
  }
}


src/c/getline.c
#line 17 "getline.nw"
static char rcsid[] = "$Id: getline.nw,v 2.21 2006/06/12 21:03:53 nr Exp nr $";
static char rcsname[] = "$Name: v2_11b $";
#define START_SIZE 128                  /* initial buffer size */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "columns.h"
#include "errors.h"
#include "getline.h"

static char *buf1 = NULL, *buf2 = NULL; /* lines without, with tabs expanded */
static buf_size = START_SIZE;           /* size of both buffers if non-NULL */

#line 42 "getline.nw"
void new_buffers(void) {
    checkptr(buf1 = (char *) realloc(buf1, buf_size));
    checkptr(buf2 = (char *) realloc(buf2, buf_size));
}
#line 47 "getline.nw"
char *getline_noweb (FILE *fp) {

    
#line 82 "getline.nw"
if (buf1==NULL) {
    checkptr(buf1 = (char *) malloc (buf_size));
    checkptr(buf2 = (char *) malloc (buf_size));
}
#line 50 "getline.nw"
    
    buf1=fgets(buf1, buf_size, fp);
    if (buf1==NULL) return buf1; /* end of file */
    while (buf1[strlen(buf1)-1] != '\n') { /* failed to get whole line */
        buf_size *= 2;
        new_buffers();
        if (fgets(buf1+strlen(buf1),buf_size-strlen(buf1),fp)==NULL)
            return buf1; /* end of file */
    }
    return buf1;
}
#line 62 "getline.nw"
char *getline_expand (FILE *fp) {
    char *s, *t;
    int width;

    if (getline_noweb(fp)==NULL) return NULL;
    
#line 87 "getline.nw"
if (columnwidth(buf1) > buf_size - 1) {
    while (columnwidth(buf1) > buf_size - 1) buf_size *= 2;
    new_buffers();
}
#line 68 "getline.nw"
    s = buf1; t = buf2; width=0;
    while (*s)
        if (*s=='\t' && tabsize > 0) {
            do {
                *t++ = ' '; width++;
            } while (width % tabsize != 0);
            s++;
        } else {
            *t++ = *s++; width++;
        }
    *t='\0';
    return buf2;
}


src/c/finduses.c
#line 5 "finduses.nw"
static char rcsid[] = "$Id: finduses.nw,v 1.19 2006/06/12 21:03:53 nr Exp nr $";
static char rcsname[] = "$Name: v2_11b $";
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "errors.h"
#include "match.h"
#include "getline.h"
#include "recognize.h"
#line 21 "finduses.nw"
static Recognizer nwindex;
#define ALPHANUM "_'@ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789#"
#define SYMBOLS "!%^&*-+:=|~<>./?`"
/* note $ and \ both delimiters */
#line 29 "finduses.nw"
static int showquotes = 1;
#line 95 "finduses.nw"
typedef struct line_and_outfile {
    char *line;
    FILE *out;
} LineOut;
#line 64 "finduses.nw"
static void read_ids(FILE *in);
#line 100 "finduses.nw"
static void add_use_markers(FILE *in, FILE *out);
#line 152 "finduses.nw"
static void write_index_use(void *closure, char *id, char *instance);
static char *emit_up_to(FILE *f, char *s, char *limit);
#line 33 "finduses.nw"
main(int argc, char **argv) {
    FILE *fp;
    int i;

    progname = argv[0];	
    for (i = 1; i < argc && argv[i][0] == '-' && argv[i][1] != 0; i++)
        if (!strcmp(argv[i], "-noquote"))
            showquotes = 0;
        else
            errormsg(Error, "%s: unknown option %s\n", progname, argv[i]);
    nwindex = new_recognizer(ALPHANUM, SYMBOLS);
    if (i == argc) {
       
#line 75 "finduses.nw"
{   FILE *tmp = tmpfile();
    char *line;
    if (tmp == NULL) 
#line 155 "finduses.nw"
errormsg(Fatal, "%s: couldn't open temporary file\n", progname);
#line 78 "finduses.nw"
    while ((line = getline_noweb(stdin)) != NULL) {
        if (fputs(line, tmp) == EOF) 
#line 157 "finduses.nw"
errormsg(Fatal, "%s: error writing temporary file\n", progname);
#line 80 "finduses.nw"
        if (is_index(line, "defn")) {
            if (line[strlen(line)-1] == '\n') line[strlen(line)-1] = 0;
            add_ident(nwindex, line+1+5+1+4+1);
        } else if (is_index(line, "localdefn")) {
            if (line[strlen(line)-1] == '\n') line[strlen(line)-1] = 0;
            add_ident(nwindex, line+1+5+1+9+1);
        } else if (is_keyword(line, "fatal")) {
	    exit(1);
	}
    }
    rewind(tmp);
    stop_adding(nwindex);
    add_use_markers(tmp, stdout);
}
#line 46 "finduses.nw"
    } else {
       
#line 56 "finduses.nw"
for (; i < argc; i++)
    if ((fp=fopen(argv[i],"r"))==NULL)
        errormsg(Error, "%s: couldn't open file %s\n", progname, argv[i]);
    else {
        read_ids(fp);
        fclose(fp);
    }
#line 48 "finduses.nw"
       stop_adding(nwindex);
       add_use_markers(stdin, stdout);
    }
    nowebexit(NULL);
    return errorlevel;          /* slay warning */
}
#line 66 "finduses.nw"
static void read_ids(FILE *in) {
    char *line;
    while ((line = getline_noweb(in)) != NULL) {
        if (line[strlen(line)-1] == '\n') line[strlen(line)-1] = 0;
        add_ident(nwindex, line);
    }
}
#line 102 "finduses.nw"
static void add_use_markers(FILE *in, FILE *out) {
    char *line;
    int incode = 0;
    LineOut info; info.line = (char *)0; info.out = out;
    
    while ((line = getline_noweb(in)) != NULL) {
        if (is_begin(line, "code") || showquotes && is_keyword(line, "quote"))
            incode = 1;
        else if (is_end(line, "code") || is_keyword(line, "endquote"))
            incode = 0;
        if (is_keyword(line, "text") && incode) {
            info.line = line + 6; /* skip "@text " */
            search_for_ident(nwindex, line, write_index_use, &info);
            if (*info.line && *info.line != '\n') 
                fprintf(out, "@text %s", info.line);    /* has newline */
        } else
            fprintf(out, "%s", line);
    }       
}
#line 133 "finduses.nw"
static void write_index_use(void *closure, char *id, char *instance) {
  LineOut *info = (LineOut *) closure;
  info->line = emit_up_to(info->out, info->line, instance);
  fprintf(info->out, "@index use %s\n", id);
  info->line = emit_up_to(info->out, info->line, instance + strlen(id));
}
#line 140 "finduses.nw"
static char *emit_up_to(FILE *f, char *s, char *limit) {
  if (s < limit) {
    char saved = *limit;
    *limit = 0;
    fprintf(f, "@text %s\n", s);
    *limit = saved;
    return limit;
  } else {
    return s;
  }
}


src/c/notangle.c
#line 18 "notangle.nw"
static char rcsid[] = "$Id: notangle.nw,v 2.22 2006/06/12 21:03:53 nr Exp nr $";
static char rcsname[] = "$Name: v2_11b $";
#define MAX_MODNAME 255
#line 23 "notangle.nw"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include "strsave.h"
#include "getline.h"
#include "modules.h"
#include "modtrees.h"
#include "errors.h"
#include "match.h"
#include "notangle.h"

#line 161 "notangle.nw"
static void warn_dots(char *modname);          /* warn about names ending in ... */

#line 180 "notangle.nw"
void insist(char *line, char *keyword, char *msg);

#line 37 "notangle.nw"
void emit_module_named (FILE *out, char *rootname, char *locformat) {
    Module root = NULL; /* ptr to root module */

    root = lookup(rootname);
    
#line 165 "notangle.nw"
if (root==NULL) {
    errormsg(Fatal, "The root module <<%s>> was not defined.", rootname);
    return;
}
#line 42 "notangle.nw"
    (void) expand(root,0,0,0,locformat,out);
    putc('\n',out);                     /* make output end with newline */
}
#line 56 "notangle.nw"
void read_defs(FILE *in) {
    char modname[MAX_MODNAME+1] = ""; /* name of module currently being read, 
                                         [[""]] if no module is being read */ 
    Module modptr = NULL;       /* ptr to current module, or NULL */
    char *line = NULL;          /* buffer for input */
    Location loc;

    while ((line = getline_noweb(in)) != NULL) {
        if (is_keyword(line, "fatal")) exit(1);
        
#line 101 "notangle.nw"
if (is_keyword(line, "nl") || is_index(line, "nl")) {
    loc.lineno++;
#line 113 "notangle.nw"
} else if (is_keyword(line,"file")) {
    
#line 126 "notangle.nw"
{ char temp[MAX_MODNAME+1];
  if (strlen(line) >= MAX_MODNAME + strlen("@file "))
    overflow("file name size");
  strcpy(temp,line+strlen("@file "));
  temp[strlen(temp)-1]='\0';
  loc.filename = strsave(temp);
}
#line 115 "notangle.nw"
    loc.lineno = 1;
} else if (is_keyword(line, "line")) {
    
#line 134 "notangle.nw"
{ char temp[MAX_MODNAME+1];
  if (strlen(line) >= MAX_MODNAME + strlen("@line "))
    overflow("file name size");
  strcpy(temp,line+strlen("@line "));
  temp[strlen(temp)-1]='\0';
  
#line 143 "notangle.nw"
{ char *p;
  for (p = temp; *p; p++)
    if (!isdigit(*p)) 
      errormsg(Error, "non-numeric line number in `@line %s'", temp);
}
#line 140 "notangle.nw"
  loc.lineno = atoi(temp);
}
#line 118 "notangle.nw"
    loc.lineno--;
#line 104 "notangle.nw"
} 
if (!is_begin(line, "code"))
    continue;
#line 66 "notangle.nw"
        
#line 97 "notangle.nw"
do { line = getline_noweb(in);
} while (line != NULL && !is_keyword(line,"defn") && !is_keyword(line,"text"));
#line 67 "notangle.nw"
        insist(line,"defn","code chunk had no definition line");
        
#line 123 "notangle.nw"
strcpy(modname,line+strlen("@defn "));
modname[strlen(modname)-1]='\0';
#line 69 "notangle.nw"
        warn_dots(modname);       /* names ending in ... aren't like web */
        modptr = insert(modname); /* find or add module in table */

        line = getline_noweb(in);
        insist(line,"nl","definition line not followed by newline");
        loc.lineno++;
        line = getline_noweb(in);
        while (line != NULL && !is_end(line,"code")) {
            if (is_keyword(line,"nl")) {
                addnewline(modptr);
                loc.lineno++;
            } else if (is_keyword(line,"text")) {
                addstring(modptr,line+1+4+1,loc);
            } else if (is_keyword(line,"use")) {
                warn_dots(line+1+3+5);
                addmodule(modptr,line+1+3+1);
            } else if (is_index(line, "nl")) {
                loc.lineno++;
            
#line 113 "notangle.nw"
} else if (is_keyword(line,"file")) {
    
#line 126 "notangle.nw"
{ char temp[MAX_MODNAME+1];
  if (strlen(line) >= MAX_MODNAME + strlen("@file "))
    overflow("file name size");
  strcpy(temp,line+strlen("@file "));
  temp[strlen(temp)-1]='\0';
  loc.filename = strsave(temp);
}
#line 115 "notangle.nw"
    loc.lineno = 1;
} else if (is_keyword(line, "line")) {
    
#line 134 "notangle.nw"
{ char temp[MAX_MODNAME+1];
  if (strlen(line) >= MAX_MODNAME + strlen("@line "))
    overflow("file name size");
  strcpy(temp,line+strlen("@line "));
  temp[strlen(temp)-1]='\0';
  
#line 143 "notangle.nw"
{ char *p;
  for (p = temp; *p; p++)
    if (!isdigit(*p)) 
      errormsg(Error, "non-numeric line number in `@line %s'", temp);
}
#line 140 "notangle.nw"
  loc.lineno = atoi(temp);
}
#line 118 "notangle.nw"
    loc.lineno--;
#line 88 "notangle.nw"
            } else if (!is_keyword(line, "index"))
                
#line 182 "notangle.nw"
errorat(loc.filename, loc.lineno, Error, "botched code chunk `%s'", line);
#line 90 "notangle.nw"
            line = getline_noweb(in);
        }
        
#line 176 "notangle.nw"
if (line==NULL) {
    impossible("End of file occurred in mid-module");
}
#line 93 "notangle.nw"
    }
}
#line 154 "notangle.nw"
static
void warn_dots(char *modname) {
  if (!strcmp(modname+strlen(modname)-3,"...")) 
    errormsg(Warning, "Module name <<%s>> isn't completed as in web", 
             modname);
}
#line 170 "notangle.nw"
void insist(char *line, char *keyword, char *msg) {
  
#line 176 "notangle.nw"
if (line==NULL) {
    impossible("End of file occurred in mid-module");
}
#line 172 "notangle.nw"
  if (!is_keyword(line,keyword))
    impossible(msg);
}
